Sources/GRPC/ReadWriteStates.swift

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
+import SwiftProtobuf
+
+/// Number of messages expected on a stream.
+enum MessageCount {
+  case one
+  case many
+}
+
+/// Encapsulates the state required to create a new write state.
+struct PendingWriteState {
+  /// The number of messages we expect to write to the stream.
+  var expectedCount: MessageCount
+
+  /// The encoding we should use when writing the message.
+  var encoding: CompressionMechanism
+
+  /// The 'content-type' being written.
+  var contentType: ContentType
+
+  func makeWriteState() -> WriteState {
+    return WriteState(
+      expectedCount: self.expectedCount,
+      writer: LengthPrefixedMessageWriter(),
+      contentType: self.contentType
+    )
+  }
+}
+
+/// The write state of a stream.
+struct WriteState {
+  /// Whether the stream may be written to.
+  internal private(set) var canWrite: Bool = true
+
+  /// The number of messages we expect to write to the stream.
+  var expectedCount: MessageCount
+
+  /// A writer to encode `Message`s into the gRPC wire-format.
+  var writer: LengthPrefixedMessageWriter
```

Comment:
Will this ever be initialized with something other than `LengthPrefixedMessageWriter()`?

Reply:
Yes: when we add compression support we'll have to initialize the writer with the compression algorithm being used.